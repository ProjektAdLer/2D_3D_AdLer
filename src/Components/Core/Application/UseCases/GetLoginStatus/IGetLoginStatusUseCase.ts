import { ISynchronousUsecase } from "../../Abstract/ISynchronousUsecase";
import LoginStatusTO from "../../DataTransferObjects/LoginStatusTO";

export default interface IGetLoginStatusUseCase
  extends ISynchronousUsecase<void, LoginStatusTO> {}
