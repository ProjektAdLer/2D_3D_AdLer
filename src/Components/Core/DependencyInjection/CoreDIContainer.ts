import { LoadWorldController } from "../Presentation/React/LoadWorldButton/LoadWorldController";
import ILoadWorldController from "../Presentation/React/LoadWorldButton/ILoadWorldController";
import { Container } from "inversify";
import CORE_TYPES from "./CoreTypes";
import ISceneView from "../Presentation/Babylon/SceneManagement/ISceneView";
import SceneView from "../Presentation/Babylon/SceneManagement/SceneView";
import SceneViewModel from "../Presentation/Babylon/SceneManagement/SceneViewModel";
import ScenePresenter from "../Presentation/Babylon/SceneManagement/ScenePresenter";
import MainScene from "../Presentation/Babylon/SceneManagement/MainScene";
import ICreateSceneClass from "../Presentation/Babylon/SceneManagement/ICreateSceneClass";
import IScenePresenter from "../Presentation/Babylon/SceneManagement/IScenePresenter";
import infrastructureDIContainer from "./InfrastructureDIContainer";
import useCaseDIContainer from "./UseCases/UseCaseDIConatiner";
import PortsDIContainer from "./Ports/PortsDIContainer";
import BuilderDIContainer from "./Builders/BuilderDIContainer";
import INavigation from "../Presentation/Babylon/Navigation/INavigation";
import Navigation from "../Presentation/Babylon/Navigation/Navigation";
import NavigationConfiguration from "../Presentation/Babylon/Navigation/NavigationConfiguration";

var CoreDIContainer = new Container();

// Scene
CoreDIContainer.bind<ISceneView>(CORE_TYPES.ISceneView)
  .to(SceneView)
  .inSingletonScope();
CoreDIContainer.bind(SceneViewModel).toSelf().inSingletonScope();
CoreDIContainer.bind<IScenePresenter>(CORE_TYPES.IScenePresenter)
  .to(ScenePresenter)
  .inSingletonScope();

// Navigation
CoreDIContainer.bind<INavigation>(CORE_TYPES.INavigation)
  .to(Navigation)
  .inSingletonScope();
CoreDIContainer.bind(NavigationConfiguration).toSelf();

// Controllers
CoreDIContainer.bind<ILoadWorldController>(CORE_TYPES.ILoadWorldController)
  .to(LoadWorldController)
  .inSingletonScope();

// bind other CreateSceneClass here for testing puposes -MK
CoreDIContainer.bind<ICreateSceneClass>(CORE_TYPES.ICreateSceneClass).to(
  MainScene
);

CoreDIContainer.load(infrastructureDIContainer);
CoreDIContainer.load(useCaseDIContainer);
CoreDIContainer.load(PortsDIContainer);
CoreDIContainer.load(BuilderDIContainer);

export default CoreDIContainer;
