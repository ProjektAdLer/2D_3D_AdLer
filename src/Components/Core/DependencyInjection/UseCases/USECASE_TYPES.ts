const USECASE_TYPES = {
  ILoadUserLearningWorldsUseCase: Symbol("ILoadUserLearningWorldsUseCase"),
  ILoadLearningWorldUseCase: Symbol("ILoadLearningWorldUseCase"),
  ILoadAvatarUseCase: Symbol("ILoadAvatarUseCase"),
  ILoadLearningElementUseCase: Symbol("ILoadLearningElementUseCase"),
  IScoreLearningElementUseCase: Symbol("IScoreLearningElementUseCase"),
  ICalculateLearningSpaceScoreUseCase: Symbol(
    "ICalculateLearningSpaceScoreUseCase"
  ),
  ICalculateLearningWorldScoreUseCase: Symbol(
    "ICalculateLearningWorldScoreUseCase"
  ),
  ILoginUseCase: Symbol("ILoginUseCase"),
  ILoadLearningSpaceUseCase: Symbol("ILoadLearningSpaceUseCase"),
  IScoreH5PLearningElementUseCase: Symbol("IScoreH5PLearningElementUseCase"),
  IGetLearningElementSourceUseCase: Symbol("IGetLearningElementSourceUseCase"),
  IGetLoginStatusUseCase: Symbol("IGetLoginStatusUseCase"),
  ISetUserLocationUseCase: Symbol("ISetUserLocationUseCase"),
  IGetUserLocationUseCase: Symbol("IGetUserLocationUseCase"),
  ICalculateLearningSpaceAvailabilityUseCase: Symbol(
    "ICalculateLearningSpaceAvailabilityUseCase"
  ),
  IGetLearningSpacePrecursorAndSuccessorUseCase: Symbol(
    "IGetLearningSpacePrecursorAndSuccessorUseCase"
  ),
};

export default USECASE_TYPES;
