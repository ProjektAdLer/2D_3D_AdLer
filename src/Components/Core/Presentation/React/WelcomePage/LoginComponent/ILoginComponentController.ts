export default interface ILoginComponentController {
  login(username: string, password: string): void;
}
