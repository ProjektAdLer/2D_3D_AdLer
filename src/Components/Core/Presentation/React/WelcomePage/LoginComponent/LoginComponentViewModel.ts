import Observable from "../../../../../../Lib/Observable";

export default class LoginComponentViewModel {
  userLoggedIn = new Observable<boolean>(false);
  modalVisible = new Observable<boolean>(true);
}
