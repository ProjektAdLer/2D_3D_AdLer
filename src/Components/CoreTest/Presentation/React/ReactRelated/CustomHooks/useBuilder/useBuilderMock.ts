import * as useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";

/**
 * Mocks the useViewModelControllerProvider hook.
 * @param mockedTupels The return value, which the mocked hook will return.
 *
 */
export default function useBuilderMock(
  mockedTupels: [viewModel: any, controller: any] = [undefined, undefined]
) {
  jest.spyOn(useBuilder, "default").mockReturnValue(mockedTupels);
}
