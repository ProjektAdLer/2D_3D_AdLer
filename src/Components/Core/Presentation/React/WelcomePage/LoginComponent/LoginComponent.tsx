import useObservable from "../../ReactRelated/CustomHooks/useObservable";
import StyledButton from "../../ReactRelated/ReactBaseComponents/StyledButton";
import LoginComponentController from "./LoginComponentController";
import LoginComponentViewModel from "./LoginComponentViewModel";
import moodleIcon from "../../../../../../Assets/icons/16-moodle/moodle-icon-nobg.svg";
import useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";
import BUILDER_TYPES from "~DependencyInjection/Builders/BUILDER_TYPES";
import { createPortal } from "react-dom";
import LoginModal from "./LoginModal";
import React, { useEffect } from "react";
import { useInjection } from "inversify-react";
import IGetLoginStatusUseCase from "src/Components/Core/Application/UseCases/GetLoginStatus/IGetLoginStatusUseCase";
import USECASE_TYPES from "~DependencyInjection/UseCases/USECASE_TYPES";

/**
 * React Component that displays a login button. When clicked, a modal will be overlayed.
 */
export default function LoginComponent() {
  const [viewModel, controller] = useBuilder<
    LoginComponentViewModel,
    LoginComponentController
  >(BUILDER_TYPES.ILoginButtonBuilder);
  const getLoginStatusUseCase = useInjection<IGetLoginStatusUseCase>(
    USECASE_TYPES.IGetLoginStatusUseCase
  );

  const [, setModalVisible] = useObservable<boolean>(viewModel?.modalVisible);
  const [userLoggedIn, setUserLoggedIn] = useObservable<boolean>(
    viewModel?.userLoggedIn
  );

  useEffect(() => {
    const loginStatus = getLoginStatusUseCase.execute();
    setUserLoggedIn(loginStatus);
    setModalVisible(!loginStatus);
  }, [getLoginStatusUseCase, setUserLoggedIn, setModalVisible]);

  if (!controller || !viewModel) return null;

  return (
    <React.Fragment>
      <StyledButton
        color={userLoggedIn ? "success" : "default"}
        onClick={() => setModalVisible(true)}
      >
        <img src={moodleIcon} alt="Moodle-Icon"></img>
      </StyledButton>
      {createPortal(
        <div className="z-10">
          <LoginModal viewModel={viewModel} controller={controller} />
        </div>,
        document.body
      )}
    </React.Fragment>
  );
}
