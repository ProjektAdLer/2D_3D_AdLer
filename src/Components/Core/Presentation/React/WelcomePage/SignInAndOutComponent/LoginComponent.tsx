import useObservable from "../../ReactRelated/CustomHooks/useObservable";
import StyledButton from "../../ReactRelated/ReactBaseComponents/StyledButton";
import SignInAndOutComponentController from "./SignInAndOutComponentController";
import SignInAndOutComponentViewModel from "./SignInAndOutComponentViewModel";
import useBuilder from "~ReactComponents/ReactRelated/CustomHooks/useBuilder";
import BUILDER_TYPES from "~DependencyInjection/Builders/BUILDER_TYPES";
import React, { useEffect, useState } from "react";
import { useInjection } from "inversify-react";
import IGetLoginStatusUseCase from "src/Components/Core/Application/UseCases/GetLoginStatus/IGetLoginStatusUseCase";
import USECASE_TYPES from "~DependencyInjection/UseCases/USECASE_TYPES";
import { AdLerUIComponent } from "src/Components/Core/Types/ReactTypes";
import tailwindMerge from "../../../Utils/TailwindMerge";
import StyledInputField from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledInputField";
import StyledPasswordField from "~ReactComponents/ReactRelated/ReactBaseComponents/StyledPasswordField";
import { useTranslation } from "react-i18next";
import ILoadAvatarConfigUseCase from "src/Components/Core/Application/UseCases/LoadAvatarConfig/ILoadAvatarConfigUseCase";

/**
 * React Component that displays a login button. When clicked, a modal will be overlayed.
 */
export default function LoginComponent({
  className,
}: Readonly<AdLerUIComponent>) {
  const [viewModel, controller] = useBuilder<
    SignInAndOutComponentViewModel,
    SignInAndOutComponentController
  >(BUILDER_TYPES.ISignInAndOutComponentBuilder);
  const getLoginStatusUseCase = useInjection<IGetLoginStatusUseCase>(
    USECASE_TYPES.IGetLoginStatusUseCase,
  );
  const loadAvatarConfigUseCase = useInjection<ILoadAvatarConfigUseCase>(
    USECASE_TYPES.ILoadAvatarConfigUseCase,
  );
  const { t: translate } = useTranslation("start");

  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");

  const handleSubmit = React.useCallback(() => {
    controller.login(userName, password);
  }, [controller, userName, password]);

  const [, setModalVisible] = useObservable<boolean>(viewModel?.modalVisible);
  const [userLoggedIn, setUserLoggedIn] = useObservable<boolean>(
    viewModel?.userLoggedIn,
  );
  const [loginFailed] = useObservable<boolean>(viewModel?.loginFailed);

  useEffect(() => {
    const loginStatus = getLoginStatusUseCase.execute();
    setUserLoggedIn(loginStatus.isLoggedIn);
    if (loginStatus.isLoggedIn) {
      const loadConfig = async () => {
        await loadAvatarConfigUseCase.executeAsync();
      };
      loadConfig();
    }
  }, [
    getLoginStatusUseCase,
    loadAvatarConfigUseCase,
    setUserLoggedIn,
    setModalVisible,
  ]);

  if (!controller || !viewModel) return null;

  return (
    <div className={tailwindMerge(className)}>
      <h2 className="col-span-6 col-start-2 row-start-2 mt-8 pt-2 text-xl font-extrabold text-center justify-self-center portrait:text-sm lg:text-xl text-adlerdarkblue portrait:row-start-1 portrait:col-start-1 portrait:col-span-8">
        {userLoggedIn
          ? translate("loginSuccessfullSubheading")
          : translate("moodleLoginSubheading")}
      </h2>

      {!userLoggedIn && (
        <div className="flex flex-col items-start justify-center">
          <form
            className="flex items-center justify-center col-span-6 col-start-2 row-start-2 gap-2 m-4 text-xl font-bold text-center rounded-lg portrait:self-end portrait:row-start-1 mobile-landscape:scale-75 portrait:scale-50 portrait:col-start-1 portrait:col-span-8"
            onSubmit={(e) => {
              e.preventDefault();
              handleSubmit();
            }}
          >
            <StyledInputField
              className="text-adlerdarkblue"
              data-testid="userName"
              placeholder={translate("userName").toString()}
              onChange={(newVal) => {
                setUserName(newVal.target.value);
              }}
            />
            <StyledPasswordField
              className="text-adlerdarkblue"
              data-testid="password"
              placeholder={translate("password").toString()}
              onChange={(newVal) => {
                setPassword(newVal.target.value);
              }}
            />
            <StyledButton
              shape="freeFloatCenter"
              data-testid="loginButton"
              onClick={handleSubmit}
            >
              <p>{translate("loginButton")}</p>
            </StyledButton>
          </form>

          {loginFailed && (
            <div className="bg-red-100 rounded-lg mobile-landscape:w-64 self-center z-50">
              <p className="m-1 text-xs font-bold text-center text-red-500">
                {viewModel.errorMessage.Value}
              </p>
              <p className="m-1 text-xs font-bold text-center text-red-500">
                {viewModel.errorMessageAdvise.Value}
              </p>
            </div>
          )}
        </div>
      )}
    </div>
  );
}
