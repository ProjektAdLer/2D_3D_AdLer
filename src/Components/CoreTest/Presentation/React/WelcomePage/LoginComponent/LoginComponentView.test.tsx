import "@testing-library/jest-dom";
import { fireEvent, render } from "@testing-library/react";
import { getByTitle, waitFor } from "@testing-library/dom";
import { Provider } from "inversify-react";
import { mock } from "jest-mock-extended";
import React from "react";
import IGetLoginStatusUseCase from "../../../../../Core/Application/UseCases/GetLoginStatus/IGetLoginStatusUseCase";
import CoreDIContainer from "../../../../../Core/DependencyInjection/CoreDIContainer";
import USECASE_TYPES from "../../../../../Core/DependencyInjection/UseCases/USECASE_TYPES";
import ILoginComponentController from "../../../../../Core/Presentation/React/WelcomePage/LoginComponent/ILoginComponentController";
import LoginComponent from "../../../../../Core/Presentation/React/WelcomePage/LoginComponent/LoginComponent";
import LoginComponentViewModel from "../../../../../Core/Presentation/React/WelcomePage/LoginComponent/LoginComponentViewModel";
import useBuilderMock from "../../ReactRelated/CustomHooks/useBuilder/useBuilderMock";

const getLoginStatusUseCaseMock = mock<IGetLoginStatusUseCase>();

describe("LoginComponent", () => {
  beforeAll(() => {
    CoreDIContainer.snapshot();
    CoreDIContainer.rebind<IGetLoginStatusUseCase>(
      USECASE_TYPES.IGetLoginStatusUseCase
    ).toConstantValue(getLoginStatusUseCaseMock);
  });

  afterAll(() => {
    CoreDIContainer.restore();
  });

  test("doesn't render without controller", () => {
    useBuilderMock([new LoginComponentViewModel(), undefined]);

    const { container } = render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );

    expect(container.firstChild).toBeNull();
  });

  test("doesn't render without view model", () => {
    useBuilderMock([undefined, mock<ILoginComponentController>()]);

    const { container } = render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );

    expect(container.firstChild).toBeNull();
  });

  test("LoginComponent calls GetLoginStatusUseCase.execute on mount", () => {
    useBuilderMock([
      new LoginComponentViewModel(),
      mock<ILoginComponentController>(),
    ]);

    render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );

    expect(getLoginStatusUseCaseMock.execute).toBeCalled();
  });

  test("LoginComponent Tailwind Styling contains normal backgroundColor if not logged in", () => {
    let fakeModel = new LoginComponentViewModel();
    fakeModel.userLoggedIn.Value = false;
    getLoginStatusUseCaseMock.execute.mockReturnValue(false);
    useBuilderMock([fakeModel, mock<ILoginComponentController>()]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );

    const buttonStyle =
      componentUnderTest.getByTestId("login-button").className;
    expect(buttonStyle).toMatchSnapshot();
  });

  test("LoginComponent Tailwind Styling contains green backgroundColor if logged in", () => {
    let fakeModel = new LoginComponentViewModel();
    fakeModel.userLoggedIn.Value = true;
    getLoginStatusUseCaseMock.execute.mockReturnValue(true);
    useBuilderMock([fakeModel, mock<ILoginComponentController>()]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );

    const buttonStyle =
      componentUnderTest.getByTestId("login-button").className;
    expect(buttonStyle).toMatchSnapshot();
  });

  test("should render modal when clicked", () => {
    useBuilderMock([
      new LoginComponentViewModel(),
      mock<ILoginComponentController>(),
    ]);

    const componentUnderTest = render(
      <Provider container={CoreDIContainer}>
        <LoginComponent />
      </Provider>
    );
    fireEvent.click(componentUnderTest.getByTestId("login-button"));

    expect(
      componentUnderTest.findByTitle("Moodle Login")
    ).resolves.toBeInTheDocument();
  });
});
