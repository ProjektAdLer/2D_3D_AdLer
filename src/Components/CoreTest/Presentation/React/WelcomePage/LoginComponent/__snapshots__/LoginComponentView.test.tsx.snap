// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent LoginComponent Tailwind Styling contains green backgroundColor if logged in 1`] = `"justify-center p-1 lg:w-16 lg:h-16 md:w-14 md:h-14 sm:w-10 sm:h-10 w-10 h-10 aspect-square flex items-center text-sm rounded-lg hover:cursor-pointer hover:text-buttonbgblue hover:bg-adlerdarkblue lg:text-xl transition ease-in-out duration-75 active:translate-x-1 active:translate-y-1 active:border-transparent text-adlerdarkblue font-regular border-t-[1px] border-l-[1px] border-b-4 border-r-4 border-adlerdarkblue overflow-hidden bg-adlergreen"`;

exports[`LoginComponent LoginComponent Tailwind Styling contains normal backgroundColor if not logged in 1`] = `"justify-center p-1 lg:w-16 lg:h-16 md:w-14 md:h-14 sm:w-10 sm:h-10 w-10 h-10 aspect-square flex items-center text-sm rounded-lg hover:cursor-pointer hover:text-buttonbgblue hover:bg-adlerdarkblue lg:text-xl transition ease-in-out duration-75 active:translate-x-1 active:translate-y-1 active:border-transparent text-adlerdarkblue font-regular border-t-[1px] border-l-[1px] border-b-4 border-r-4 border-adlerdarkblue overflow-hidden bg-buttonbgblue"`;
